### PR TITLE
Feature/move content

### LINKED ISSUE #136: Move Content Live
In order to have content available in the `live` database for APIs, it needs to be moved there

**Definitions**
* Now - The current date and time
* Live - Publicly available in the API and the `live` database the API draws from
* Sunrise - The date, time, and timezone content is scheduled to go live
* Sunset - The date, time, and timezone content is schedule to stop being live
* Approved- Content that has gone through all steps in its [workflow](https://github.com/punchcard-cms/punchcard/pull/105)

```gherkin
Feature: Publish Content
  As a content creator
  I want my approved revisions to be scheduled to go live and come down
  So that the most up-to-date and accurate 

  Scenario: Schedule Sunrise
    Given that a content revision I have created has been approved
      And that its sunrise is later than now
     When now is equal to the revision's sunrise
     Then it should replace the live revision of its content
      And if there are multiple approved revisions with the same sunrise, only the most recently approved should be made live

  Scenario: Immediate Sunrise
    Given that a content revision I have created has a sunrise
     When the content revision is approved
      And its sunrise is now or earlier
     Then it should replace the live revision

  Scenario: Schedule Sunset
    Given that a content revision I have created has been approved
      And that it has a sunset
      And that its sunset is later than now
     When now is equal to the revision's sunset
     Then the content it is a revision for should be removed from being live

  Scenario: Change Sunset
    Given that a content revision I have created has a sunset
      And the current live revision of its content also has a sunset
     When the content revision is approved
     Then the content it is a revision of should have its sunset changed to the approved revision's sunset

  Scenario: Cancel Sunset
    Given that a content revision I have created does not have a sunset
      And the current live revision of its content does
     When the content revision is approved
     Then the content it is a revision of should have its sunset removed

  Scenario: Immediate Sunset
    Given that a content revision I have created has a sunset
     When the content revision is approved
      And its sunset is now or earlier
     Then the content it is a revision for should be removed from being live


Feature: Approve Content
  As a content approver
  I want to understand how the content I'm reviewing is going to be affected by sunrise/sunset
  So that I can make an informed decision about whether or not I should approve or reject the revision being reviewed

  Scenario: Cancel Sunset
    Given that I am reviewing a content revision without a sunset
     When that revision's content has an existing live revision
      And that the live content has a sunset
      And the sunset has not yet occurred
     Then I should be notified on the approval page that approving this revision will cancel the existing sunset
      And I should be notified of the existing sunset

  Scenario: Change Sunset
    Given that I am reviewing a content revision with a sunset
      And that revision's sunset is past now
     When that revision's content has an existing live revision
      And that the live content has a sunset
      And the sunset has not yet occurred
     Then I should be notified on the approval page that this revision will change the existing sunset
      And I should be notified of the extent of the change (how much further/closer to now the change is)

  Scenario: Immediate Sunset
    Given that I am reviewing a content revision with a sunset
      And that revision's sunset is before now
     When that revision's content has an existing live revision
      And that the live content has a sunset
      And the sunset has not yet occurred
     Then I should be notified on the approval page that this revision will immediately sunset its content 

  Scenario: Sunrise That Which Is Sunset
    Given that I am reviewing a content revision with a sunrise
     When that revision's content has an existing live revision
      And that the live content has a sunset
      And the sunset has already occurred
     Then I should be notified on the approval page that this revision will sunrise the currently sunset content

  Scenario: Sunrise That Which Will Be Sunset
    Given that I am reviewing a content revision with a sunrise
     When that revision's content has an existing live revision
      And that the live content has a sunset
      And the sunset has not yet occurred
      And the sunset will occur before the current revision's sunrise
     Then I should be notified on the approval page that this revision will sunrise the content that otherwise would be sunset
      And I should be notified of the difference between when the content would be scheduled to sunset and sunrise

  Scenario: Immediate Sunrise
    Given that I am reviewing a content revision with a sunrise
     When that revision's sunrise is equal to or before now
     Then I should be notified on the approval page that this revision will be live immediately after approval

  Scenario: Sunrise Override
    Given that I am reviewing a content revision with a sunrise
     When that revision's content has existing revisions scheduled for sunrise
      And this revision's sunrise is the same as other revisions that are scheduled for sunrise
     Then I should be notified on the approval page that this revision will supersede the other scheduled revisions with the same sunrise

  Scenario: View Sunrise and Sunset
    Given that I am reviewing a content revision
     When I want to check the sunrise or sunset value of the revision being reviewed
     Then I should be able to readily see them just as I can the other fields of a revision

  Scenario: View Revision Schedule
    Given that a piece of content can have multiple revisions scheduled to sunrise at once
      And a piece of content may be scheduled to sunset
      And a piece of content my have a live revision
     When I want to understand how a piece of content may change over time
      And I can see what revision is current live, if any
     Then I can see a listing of approved revisions for a piece of content
      And I can see the revision ID, date created, date approved, identifier, have a link to view the complete revision, and have a link to create a new revision off of the current one, for each revision
      And if the revision has not yet passed its sunrise, its sunrise date and time until sunrise
      And revisions should be sorted by when they will become live
      And the current live revision should be separated from all other revisions
      And the sunset date for the content should be visible
      And if there are revisions scheduled to become live after the sunset date, they should be separated from revisions scheduled to become live before the sunset date.
      And I should be able to access this view from the approval page
      And I should be able to access this view from the revision listing page
      And I should be able to access this from the content listing page
```

---

**Publish Content**
- [ ] Schedule Sunrise
- [ ] Immediate Sunrise
- [ ] Schedule Sunset
- [ ] Change Sunset
- [ ] Cancel Sunset
- [ ] Immediate Sunset

**Approve Content**
- [ ] Cancel Sunset
- [ ] Change Sunset
- [ ] Immediate Sunset
- [ ] Sunrise That Which Is Sunset
- [ ] Sunrise That Which Will Be Sunset
- [ ] Immediate Sunrise
- [ ] Sunrise Override
- [ ] View Sunrise and Sunset
- [ ] View Revision Schedule

### PR DESCRIPTION
**Pull request by Snugug**
Schedules content to go live.

Takes precedence over #352

---
Resolves #136
Resolves #352

`DCO 1.1 Signed-off-by: Sam Richard <sam@snug.ug>`

### CODE REVIEW

**Comment by Snugug** [comment]
Tests not written yet

**Comment by Snugug** [comment]
@scottnath I'll start writing tests Tuesday. You can look at the code whenever

File: lib/schedule/index.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+const Promise = require('bluebird');
+const _ = require('lodash');
+
+const utils = require('./utils');
+const database = require('../database');
+
+/*
+ * Schedule content sunrise/sunset
+ */
```
Comment (scottnath) [inline lib/schedule/index.js]: please add jsdoc

**Comment by scottnath** [comment]
@Snugug no issues - except a need for JSDoc, couple q's:

1. What is the issue with rejection of promises not working?
2. With `node-schedule`, what if the app re-boots? I think all the jobs are re-scheduled by bootstrap - correct?

**Comment by Snugug** [comment]
- We can't properly reject the errors in our scheduling without blocking our thread, so they're just logged instead
- Bootstrap will solve all of that

**Pull request merged**